raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java

```diff
@@ -272,20 +280,109 @@ public void testUpdateHighWatermarkQuorumSizeThree() {
         assertEquals(Optional.empty(), state.highWatermark());
         assertTrue(state.updateReplicaState(node2, 0, new LogOffsetMetadata(15L)));
         assertEquals(Optional.of(new LogOffsetMetadata(15L)), state.highWatermark());
-        assertFalse(state.updateLocalState(new LogOffsetMetadata(20L)));
+        assertFalse(state.updateLocalState(new LogOffsetMetadata(20L), voterSet));
         assertEquals(Optional.of(new LogOffsetMetadata(15L)), state.highWatermark());
         assertTrue(state.updateReplicaState(node1, 0, new LogOffsetMetadata(20L)));
         assertEquals(Optional.of(new LogOffsetMetadata(20L)), state.highWatermark());
         assertFalse(state.updateReplicaState(node2, 0, new LogOffsetMetadata(20L)));
         assertEquals(Optional.of(new LogOffsetMetadata(20L)), state.highWatermark());
     }
 
+    @Test
+    public void testUpdateHighWatermarkAddingFollowerToVoterStates() {
```

Comment:
Can we add a test for the decreasing case? For example this change should not decrease the HWM.
```text
voter 1 LEO: 15
voter 2 LEO: 15
voter 3 LEO: 10
```

The HWM is 15.
Adding voter 4 should not decrease the HWM to 10

```text
voter 1 LEO: 15
voter 2 LEO: 15
voter 3 LEO: 10
voter 4 LEO: 5
```

The HWM should be 15 even though the naive computation would return 10.